feat: design fine grained access control rule engine (#1790)
mia-platform/documentation: docs/development_suite/api-console/api-design/limit-user-actions.md

> # Granular role-based access control

Reviewer: ```suggestion
# Design changes Fine-Grained Access Control
```

> This feature introduces a mechanism for managing user interactions within the application by restricting specific actions based on user roles and specified rules.

Reviewer: ```suggestion
This feature introduces a mechanism for managing user interactions within the application by restricting specific actions based on user roles and specified rules in the Design section.
```

> If a user try to save a configuration after performing some changes that are forbidden by the configured the rules, the saving is blocked and an error is returned.

Reviewer: ```suggestion
If a user attempts to save a configuration after performing some changes that are forbidden by the configured rules, the saving is blocked and an error is returned.
```

*Rules* and *user roles* can be configured at both the Project and the Company levels.
The following logic is applied:

> - **Rules**: project and company rules are combined together without conflict, as only disallow rules are used.  

Reviewer: Project and Company always with the first Capital letter

```suggestion
- **Rules**: Project and Company rules are combined together without conflict, as only disallow rules are used.  
```

> - **User Roles**: If a user has roles assigned at the project level, only those roles are applied. If no project-level roles are assigned, the roles defined at the company level are used instead.

Reviewer: ```suggestion
- **User Roles**: If a user has roles assigned at the project level, only those roles used to find the applicable rules. If no project-level roles are assigned, the roles defined at the company level are used instead.
```

| ------------------------- | --------------------------------------------------------------------------------------------------- |
| `endpoints.security.edit` | block edit of the fields [`public`, `acl`, `secreted`] of `endpoints` and `routes` inside endpoints |
|                           |                                                                                                     |

## Configuring Rules Via API

### Updating rules on a Project

The API for updating the rules on a Project is defined as follows

> **NOTE**  
> This API is meant for internal use and will be subject to breaking changes.
>

#### Request

- verb: `PATCH`
- path: `/api/backend/projects/:projectId/rules`

**Authentication required**

##### Body

The **body** of the request has the structure described in [Configuration definition](#configuration-definition)

### Updating rules on a Company

The API for updating the rules on a Companty is defined as follows.

> **NOTE**  
> This API is meant for internal use and will be subject to breaking changes.
>

#### Request

- verb: `PATCH`
- path: `/api/backend/tenants/:tenantId/rules`

> **Authentication required**

Author: Here should I add also the user permission needed to call this API?

Reviewer: yes

Author: done

##### Body

The **body** of the request the structure described in [Configuration definition](#configuration-definition)

> ### Examples

Reviewer 2: ```suggestion
### Use Case Examples
```

Here are some examples of request bodies for updating Project or Company rules (the body is the same for the two API)

> Prevent edit of the `dockerImage` of all services to the role `maintainer`

Reviewer 2: ```suggestion
#### Prevent edit of the `dockerImage` of all services to the role `maintainer`
```

                "disallowedRuleSet": [
                    {
                        "jsonPath": "$.services.*.dockerImage"
                    },
                ],
                "roleIds": [
                    "maintainer"
                ]
            }
        ]
    }
}
```

Prevent creation of the resource `secrets` to the role `maintainer`

```json
{
  "configurationManagement": {
    "saveChangesRules": [
      {
        "disallowedRuleSet": [
          {
            "jsonPath": "$.collections",
            "processingOptions": {
              "action": "create"
            }
          }
        ],
        "roleIds": ["maintainer"]
      }
    ]
  }
}
```

Through jsonpath syntax, more complex rules can be configured. The following rule for example prevent the creation of a services of a specific type (`custom-resource`) to the role `maintainer`

```json
{
  "configurationManagement": {
    "saveChangesRules": [
      {
        "disallowedRuleSet": [
          {
            "jsonPath": "$.services.[?(@.type==\"cursom-resource\")]",
            "processingOptions": { "action": "create" }
          }
        ],
        "roleIds": ["maintainer"]
      }
    ]
  }
}
```

> Configure the predefined rule with `ruleId` "endpoints.security.edit" to the role `maintainer`

Reviewer 2: ```suggestion
#### Configure the predefined rule with `ruleId` "endpoints.security.edit" to the role `maintainer`
```

---
id: limit-user-actions
title: Limit User Actions

> sidebar_label: Limit User Actions

Author: I used "Fine-Grained Access Control"

          ]
      }
  }
```

- prevent creation of the resource `secrets` to the role `maintainer`

```json
{
  "configurationManagement": {
    "saveChangesRules": [
      {
        "disallowedRuleSet": [
          {
            "jsonPath": "$.collections",
            "processingOptions": {
              "action": "create"
            }
          }
        ],
        "roleIds": ["maintainer"]
      }
    ]
  }
}
```

- prevent the creation of a services of a specific type (`custom-resource`) to the role `maintainer`

```json
{
  "configurationManagement": {
    "saveChangesRules": [
      {
        "disallowedRuleSet": [
          {
            "jsonPath": "$.services.[?(@.type==\"cursom-resource\")]",
            "processingOptions": { "action": "create" }
          }
        ],
        "roleIds": ["maintainer"]
      }
    ]
  }
}
```

- configure the predefined rule with `ruleId` "endpoints.security.edit" to the role `maintainer`

```json
{
  "configurationManagement": {
    "saveChangesRules": [
      {
        "disallowedRuleSet": [
          {
            "ruleId": "endpoints.security.edit"
          }
        ],
        "roleIds": ["maintainer"]
      }
    ]
  }
}
```

> ## Fetching and configuring rules via `miactl`

Reviewer: let's put miactl before APIs

Reviewer: We should also add an info that the miactl commands are available from miactl v0.16.0 (which is yet to be released)

> ### list

Reviewer: command documentation will be aggregated in this repo (e.g. https://docs.mia-platform.eu/docs/cli/miactl/commands) once the miactl PR https://github.com/mia-platform/miactl/pull/228 will be merged, so we can link it instead of repeating it here.

I'd just just give the 4 command lines + file examples here

```
miactl company rules list --company-id=my-company
miactl company rules list --company-id=my-company --project-id=my-project
miactl company rules update --company-id=my-company  -f ~/my-rules.json
miactl company rules update --company-id=my-company --project-id=my-project  -f ~/my-rules.json
```
